### PR TITLE
feat: small changes for bucket method (Proof-642)

### PR DESCRIPTION
**Pull request by rnburn**
# Rationale for this change

Small additions and improvements made for bucket method.

# What changes are included in this PR?

* Use a fixed random seed for benchmark so that results are reproducible
* Add additional memory utility for copying to device
* Add a concurrent_for_each for index_range iterators

# Are these changes tested?

Yes.

### CODE REVIEW

File: sxt/execution/device/for_each.cc [hunk 1]
```diff
@@ -30,12 +30,11 @@ namespace sxt::xendv {
 // concurrent_for_each
 //--------------------------------------------------------------------------------------------------
 xena::future<>
-concurrent_for_each(basit::index_range rng,
+concurrent_for_each(basit::index_range_iterator first, basit::index_range_iterator last,
```
Comment (tlovell-sxt) [inline sxt/execution/device/for_each.cc:33]: Correct me if I'm wrong - this basically allows you to define a custom index_range_iterator whose step isn't bound to the number of cuda devices, right? But no matter what, `last` has to be the final value of the index_range_iterator once it's exhausted?

Would it make more sense for the parameters to just be an `index_range` and a step value to pass to `basit::split`, then the other definition passes in `basdv::get_num_devices()` as the step value? It seems like the `last` parameter would make this error if defined wrong. Or am I missing something?

File: sxt/execution/device/for_each.cc [hunk 2]
```diff
@@ -44,4 +43,11 @@ concurrent_for_each(basit::index_range rng,
     co_await std::move(fut);
   }
 }
+
+xena::future<>
+concurrent_for_each(basit::index_range rng,
```
Comment (tlovell-sxt) [inline sxt/execution/device/for_each.cc:48]: Should this have its own doc comment?

**Review by tlovell-sxt — COMMENTED** [review]
Took me a while to read this since I have very little c++ experience, thanks for your patience

File: sxt/execution/device/for_each.cc (hunk 1 above)
Comment (rnburn) [reply sxt/execution/device/for_each.cc:33]: Before calling the function, the code I'm working on needs to allocate some memory that's dependent on the distance (last - first); (ie the number of chunks). With this version, it has the first and last readily available to figure out the chunk count.

Also, passing in iterators is pretty conventional for C++ functions, see for example standard library functions: https://en.cppreference.com/w/cpp/algorithm/find

File: sxt/execution/device/for_each.cc (hunk 2 above)
Comment (rnburn) [reply sxt/execution/device/for_each.cc:48]: Added some documentation to the header file.

File: sxt/execution/device/for_each.cc (hunk 1 above)
Comment (tlovell-sxt) [reply sxt/execution/device/for_each.cc:33]: Makes sense. Thanks for the info

**Review by tlovell-sxt — APPROVED** [review]
Thanks for the extra docs, LGTM

**Pull request merged**